### ISSUE TITLE
Add preference: show minimized windows at the end of the list

### ISSUE THREAD

**Issue by mfn** [lwouis/alt-tab-macos#289]
One more default behaviour from HyperSwitch I sorely miss: minimized windows still appear in their "chronological window order".

Here is the "window setup", which involves at least 3 windows to make sense; my key combination is "cmd-tab"
- have focus on window "A"
- cmd-tab
- have focus on window "B"
- cmd-tab + tab
- have focus on another window, "C"
- hide window "C"
- cmd-tab
- have focus on window "B"

AltTab behaviour:
- cmd-tab
- have focus on window "C" => the minimized window will be resurrected 😢 

HyperSwitch behaviour:
- cmd-tab
- have focus on window "A" 😃 

HyperSwitch "moves" minimized windows to the _end_ of the list of all open windows and takes it out where it was, allowing "alt tabbing" between multiple windows to change the pattern in which they're accessed.

To me this makes very much sense, to me it's hiding a window is like saying: "go away, I'll come back _later_". Otherwise I wouldn't hide it.

Yet in AltTab, it stays where it is. It's visually marked but still keeps it's order. To me this is very undesired, as I constantly "bring back" windows I actually wanted to "get out of my way".

**Comment by lwouis** [comment]
Hi @mfn! Thanks for the feedback!

First I would like to clarify HyperSwitch behavior, because I couldn't understand/reproduce what you describe. Hidden windows are windows of a hidden app. You can't hide just 1 window of an app for instance. Only the whole app and thus all its windows. If you hide an app, HyperSwitch doesn't show any of its window, unless I'm missing something?

I think you may be saying "hidden" but meaning "minimized" perhaps. Windows can be minimized into the Dock 1 at a time. HyperSwitch doesn't show minimized windows either with the main shortcut (`alt-tab`), however it does when doing the "Active app" shortcut (`alt-backtick`). This means that you can only access them by first switching to their app, then hitting that secondary shortcut. That's not a great UX I think. Easy to forget that you have minimized windows, then what do you do, you go app by app until you find the one that has minimized windows. I think AltTab does a way better job by having a checkbox to decide or not to display minimized or hidden windows overall.

If I'm guessing right that you meant to discuss minimized windows, and you would like them to be displayed at the end of the list, then I have some issues in terms of UX.

AltTab stands out from all other similar app by having a chronological ordering instead of a z-index ordering. This was implemented in #24 and I think was a big step forwards. The app acts like Windows 10, or Android, where you remember how back in the past your last interaction with a window was, and you hit `tab` as many times as it was far back, to reach that old window.

Also your workflow of minimizing windows to put them both out of sight, and you desire putting them also far away in time artificially is not the only way think about minimizing. I think most people want the chronology to be kept accurate. This is again what Windows 10 and Android do.

This means if would be a preference. Something like "show minimized/hidden windows at the end". Another way to handle your workflow would be to change the current "Show minimized/hidden window" checkboxes to be per-shortcut. That's what I designed in my original redesign of the preferences. See the screenshots/mockups at the end of #72. This is something I would like to do eventually, but adding dynamic shortcuts is heavy work and makes the preferences a more complex UX for everyone, even though I think most users are satisfied with the current preferences, based on the feedback I'm seeing in the issues.

**Comment by mfn** [comment]
Meh, you are right. I was "over-focused" I guess => this whole thing is about "minimizing" (cmd-m). No idea how I managed to mix up the essential term here => to de-confuse the future, I adapted the title/description!

**Comment by mfn** [comment]
> HyperSwitch doesn't show minimized windows either with the main shortcut (`alt-tab`),

Yes, I so much love this :)

- it gets out of the way on the main window screen
- yet I can still access via the "Active app" shortcut

I keep forgetting that AltTab does not support the latter as I'm still running HyperSwitch for that in parallel :-(

**Comment by lwouis** [comment]
> Another way to handle your workflow would be to change the current "Show minimized/hidden window" checkboxes to be per-shortcut. That's what I designed in my original redesign of the preferences. See the screenshots/mockups at the end of #72. This is something I would like to do eventually, but adding dynamic shortcuts is heavy work and makes the preferences a more complex UX for everyone, even though I think most users are satisfied with the current preferences, based on the feedback I'm seeing in the issues.

I feel like this design would satisfy you, and is more flexible/rich that the specific checkbox to put minimized windows at the end of the list.

If that's the case then maybe we should close this ticket and reference it in a new ticket that focuses on the multiple shortcuts enhancement.

**Comment by mfn** [comment]
Are you referring to e.g. https://github.com/lwouis/alt-tab-macos/issues/72#issuecomment-585218543

Sorry, I'm not sure I follow. What is your suggestion? (apologies)

**Comment by lwouis** [comment]
Hey @mfn! I think the recent additions to AltTab are covering the issues you listed in this thread. Indeed, it is now possible to have 2 shortcuts with each their own rules, such as not showing minimized windows, or showing only the active app.

That being said, I was looking at Contexts today, for #384, and really liked their idea for minimized windows:

![image](https://user-images.githubusercontent.com/106195/89564672-51d72f80-d858-11ea-91e4-3e9c53a77891.png)

Btw, I'm super happy to note that the Contexts guys came up with the same strategy as I did for the shortcuts. Feels good to get validated on this design. I spent a *lot* of time thinking about a UI that explains the situation as well as let users have insane flexibility in the workflows they want to have:

![image](https://user-images.githubusercontent.com/106195/89564950-c01bf200-d858-11ea-9fdb-5b83021101e6.png)

I think I will implement this as a replacement for the binary show/hide checkboxes. It's added flexibility for users.
